# STAND: Sequence Taxonomy Assignment from the NCBI Database

This repository contains a handful of scripts designed to assist in assigning taxonomy to metabarcoding sequences. These automate the download of sequence and taxonomy information from the NCBI, construct a local database of sequences (and their accompanying taxonomies) that could be amplified by user-specified primers, and assign taxonomy to ASVs through BLAST+ searches.

## Requirements

POSIX-compliant operating system\
Python 3.X\
Biopython 1.7X\
NCBI Entrez Programming Utilities (E-Utilities)\
Cutadapt 2.X\
NCBI BLAST+ 2.X

More detailed documentation can be provided by running each script with the -h or --help arguments. What follows is a brief tutorial on assigning taxonomy to sequences generated using the *trn*L *g* and *h* primers.

## Downloading sequences and taxonomy from the NCBI database

Downloading of sequence and taxonomy data is performed with the download_data.py script. This script requires users to specify their query in the format used by the NCBIâ€™s Nucleotide portal (https://www.ncbi.nlm.nih.gov/nucleotide/). For example, the following command will download all chloroplast nucleotide sequences and taxonomies from the NCBI:

`download_data.py -q "biomol_genomic[PROP] AND is_nuccore[filter] AND chloroplast[filter]" -o NCBI.Chloroplast.Sequences.2021.04.05.fasta`

Depending on the number of sequences requested, this step can take a considerable amount of time.

## Building a local database for taxonomy assignment

Once the sequences and taxonomies are downloaded, the next step is to build a database with the build_database.py script. This program requires the FASTA produced from the download_data.py script, as well as the primers used for barcoding. If desired, it can restrict the final database to a subset of taxonomic ranks (due to the detailed records provided by the NCBI, restriction to only the ranks of interest is highly recommended).

An example command to generate a database for Viridiplantae *trn*L sequences, requiring no more than 3 total primer mismatches (with a maximum of 2 in either primer) and an amplified sequence length between 8 and 175 bp is as follows:

`build_database.py -f NCBI.Chloroplast.Sequences.2021.04.05.fasta -5 GGGCAATCCTGAGCCAA -3 CCATTGAGTCTCTGCACCTATC -p blast --mismatches 3 --mismatches_5 2 --mismatches_3 2 --min 8 --max 175 --require Viridiplantae --exclude environmental_samples --ranks Kingdom Phylum Subphylum Order Family Genus Species -o NCBI.trnL.Database.2021.04.05.fasta`

If desired, a database compatible with DADA2 taxonomic assignment (though not with the BLAST+ assignment script included in this repository) may be generated with the -p dada2 argument.

## Assigning taxonomy to ASVs

Once the local database is constructed, ASV taxonomy may be assigned using the classify_sequences.py script. This script requires the FASTA generated by the build_database.py script, as well as a table of DADA2 ASV counts in the CSV format. The output generated by this script is a CSV containing taxonomic assignments for each ASV, designed for easy import into R.

This script can also take an optional text file of taxonomic ranks that should receive preferential assignment if (and only if) an ASV has equally scoring best hits to multiple taxa; this file should be specified with the --taxa argument. In the case of the *trn*L data, this can be quite useful as multiple plant species from different geographic areas may yield the same ASV. Since we know which plants should be present in the areas our samples were collected from, we can tell the script to take this into account when assigning taxonomy. For such preferential assignment, the script requires a plain text file with one entry per line containing taxonomic information in the following format (Rank:Value):

```
Genus:Abies
Genus:Abronia
Genus:Abutilon
Genus:Acacia
```

An example command to run this script, assign taxonomy at each rank with the requirement that 50% of the best hits share that taxonomic rank, and preferentially assign taxonomy to genera we know are present in the area is as follows:

`classify_sequences.py -c seqtab_all_trnL_Cutadapt_nochim_17July19.csv -d NCBI.trnL.Database.2021.04.05.fasta --taxa USDA.Genera_Present_in_Area.txt --consensus 50 -o trnL.Consensus_50.2021.04.05.csv`
